Improve HHE909 error message to show incompatible files Fixes #7748

sanskar717 opened the pull request:
## Description

Improves the HHE909 error message to show which files have incompatible Solidity version requirements, making it easier to diagnose version compatibility issues in projects with large dependency trees.

## Problem

When a file can't be compiled due to version pragma incompatibilities deep in its dependency tree, the error message only says "No solc version enabled in this profile is compatible with this file and all of its dependencies" - without telling you *which* files are actually causing the problem.

This makes debugging extremely difficult, especially in projects with large dependency trees where the incompatible pragma could be buried several levels down.

## Solution

The improved error message now displays:
- ✅ Which specific files have incompatible version requirements
- ✅ The exact version pragma for each incompatible file
- ✅ Available compiler versions from the config
- ✅ A count of remaining files if there are more than 5 (to keep output manageable)

## Changes Made

- Added detailed file information to HHE909 error output
- Shows version pragma for each incompatible file
- Displays available compiler versions for reference
- Limits display to first 5 files with indicator of total count
- Improves developer experience when troubleshooting Solidity version issues

## Related Issue

Fixes #7748

## Testing

Tested with projects having:
- Single incompatible file in dependency tree
- Multiple incompatible files
- Large dependency trees (verified output is bounded to 5 files)

## Checklist

- [x] Changes follow the existing code style
- [x] Tests added/updated (if applicable)
- [x] Documentation updated (if applicable)
- [x] No breaking changes

sanskar717 committed:
Improve HHE909 error message to show incompatible files Fixes #7748

sanskar717 commented:
## Summary

This PR improves the HHE909 error message by showing which specific files have incompatible Solidity version requirements, making debugging much easier.

## Testing Evidence

Here's the improved error output in action:

[
<img width="1635" height="764" alt="image" src="https://github.com/user-attachments/assets/0e84a2d3-2631-4e4d-a487-8618bc544606" />
]

### Before vs After

**Before:** ❌ Only showed "No solc version enabled" without details
**After:** ✅ Shows exactly which file (`./contracts/IFoo.sol`) requires what version (`0.8.28`)

### Why This Helps

- Developers immediately see the problematic file
- No need to manually trace dependency trees
- Available versions clearly listed for reference
- Perfect for large projects with complex dependencies

## Testing Verified

- ✅ Single incompatible file in dependency tree
- ✅ Multiple incompatible files  
- ✅ Large dependency trees (output capped to 5 files)

kanej commented on v-next/example-project/hardhat.config.ts.save:
```diff
@@ -0,0 +1,178 @@
+port { defineConfig } from "hardhat/config";
```
Yes, I don't seem to have push permissions to your branch.

kanej replied:
Can you add a changeset as well?

kanej replied:
That would be great, thanks!

sanskar717 replied:
@kanej Changeset added! Thank you for your guidance and patience throughout this process. Looking forward to your review!

sanskar717 committed:
Remove accidentally included file

Removed hardhat.config.ts.save file that was included by mistake as pointed out in the review.

kanej commented on v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/solc-config-selection.ts:
```diff
@@ -154,12 +154,66 @@ export class SolcConfigSelector {
+    const incompatibleFiles: Array<{ path: string; pragma: string }> = [];
+
+    // check root file
+    if (maxSatisfying(compilerVersions, rootVersionRange) == null) {
```
```suggestion
    if (maxSatisfying(compilerVersions, rootVersionRange) === null) {
```
This is to satisfy our linter.

sanskar717 committed:
Update v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/solc-config-selection.ts

Co-authored-by: John Kane <john@kanej.me>

sanskar717 committed:
Update v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/solc-config-selection.ts

Co-authored-by: John Kane <john@kanej.me>

sanskar717 committed:
Add changeset
